### DISCUSSION TITLE
user.getIdToken() Throwing Error

### DISCUSSION THREAD

**Discussion by JoeToeniskoetter** [invertase/react-native-firebase#5990] [Q&A]
Hello, 

Having some issues extracting the id token from the firebase user. 
For example, when trying to extract the token immediately after a user logs in, the token is undefined.

Edit: This is only happening on my physical device. Simulator is working correctly.

```
 React.useEffect(() => {
    const unsubscribe = auth().onAuthStateChanged((user) => {
      if (user) {
        console.log("USER SIGNED IN: ", user)
        setUser(user);
        user.getIdToken().then(token => console.log("TOKEN: ", token)) //undefined
      } else {
        setUser(null);
      }
    });

    if (!auth().currentUser) {
      auth().signInAnonymously();
    }


    return unsubscribe;
  }, []);
```

and if I try to extract the token deeper in my component tree, I'm receiving an error  `[Error: [auth/internal-error] An internal error has occurred, please try again.]`

```
const MyComponent:React.FC<{}> = () => {
   const {user} = useFirebaseUser();  //from custom context

   React.useEffect(()=>{
       if(user){
            user.getIdToken().then(token => console.log(token)); //error
        }
       
   },[user])

}
```

Am I doing something incorrectly here?

Thanks,
Joe

**Comment by mikehardy** [comment]
Hi there :wave: - this looks like an issue more than a discussion, but opening it as a discussion avoids the issue template which would allow me to do a quick scan to make sure all the dependencies / versions in use are up to date. So you'll have to attest to that - are you using the latest / current stable releases of all the things in play here? Hopefully so.

It does not appear that you are doing anything wrong, but typically an `auth/internal-error` will show something interesting in the console if you launch from Xcode and watch it (or you have device plugged in and see it scroll by via Console.app), usually an exception with a stack trace and a more interesting message. Knowing the content of that will probably be illuminating.

The API documents token as nullable, though does not describe the conditions where it would be null - I think it's an async process so perhaps the token is still being minted and it will be available later? unknown. https://firebase.google.com/docs/reference/ios/firebaseauth/api/reference/Classes/FIRUser#-getidtokenresultwithcompletion:

I was about to dig in and see if our e2e probed this case, then your mention of simulator working / real device not working gave me pause.

Could even be device specific? Opens up a bunch more questions about reproducibility etc

https://stackoverflow.com/help/how-to-ask
https://stackoverflow.com/help/minimal-reproducible-example

I can only assume since you used "Simulator" vs "Emulator" that this is on iOS? Does it happen on Android?
Does this work on Simulator emulator but fail on device (which device?) if you make a clean + minimal App.js and drop it in a clean + minimal RNFB scaffold like https://github.com/mikehardy/rnfbdemo/blob/main/make-demo.sh ?

**Reply by JoeToeniskoetter** [reply]
@mikehardy I know this was quite a while ago but figured I would give an update in case this sparks any ideas. I've found that if I create a new bundle ID, generate a new GoogleService-Info.plist file, then rebuild the app, all firebase auth functionality works correctly on my real device. There must be something strange going on with that particular bundle id on my real device. Simulators still have no issue for either bundle id. I rebuilt the app using swiftui and got a slightly more informative error message 


```
Error Domain=FIRAuthErrorDomain Code=17999 "An internal error has occurred, print and inspect the error details for more information." UserInfo={FIRAuthErrorUserInfoNameKey=ERROR_INTERNAL_ERROR, NSLocalizedDescription=An internal error has occurred, print and inspect the error details for more information., NSUnderlyingError=0x282470ff0 {Error Domain=FIRAuthInternalErrorDomain Code=3 "(null)" UserInfo={NSUnderlyingError=0x282491f50 {Error Domain=com.google.HTTPStatus Code=400 "(null)" UserInfo={data={length = 111, bytes = 0x7b0a2020 22657272 6f72223a 207b0a20 ... 220a2020 7d0a7d0a }, data_content_type=application/json; charset=UTF-8}}, FIRAuthErrorUserInfoDeserializedResponseKey={
    code = 400;
    message = "INVALID_REFRESH_TOKEN";
    status = "INVALID_ARGUMENT";
}}}}
```

**Reply by mikehardy** [reply]
This is not really useful but you're on to something.
The SDKs use the REST APIs under the covers and this is apparently a common error https://firebase.google.com/docs/reference/rest/auth#section-refresh-token

Why would the refresh token be invalid? Why only for that project? *maybe* there is a restriction in the Google Cloud Platform for the keys related to that bundle-id / project such that they are not able to get refresh tokens, thus cannot exchange them for identity tokens so nothing works :thinking: - this is commentary at quality/confidence level of "thinking out loud", but it's the only ideas I've got, maybe useful, sincere apologies if not

**Comment by JoeToeniskoetter** [comment]
@mikehardy Finally figured it out....

Saw a log in xcode showing the token stored on my device expired on 11/23/2021 which was strange..

Called `auth().signOut()` on start up, relaunched the app and it was able to get a new token. I guess there was a stale refresh token saved to my device and apparently associated with the bundle ID. Since the refresh token was expired, it wasn't able to exchange it for a new access token. Once I cleared out the old token via `auth().signOut()`, it was able to authenticate anonymously again and get a new token.

Even more frustrating, I bought a new iPhone last year and this token was even apparently transferred from the old phone since this issue was happening on the old one as well. Yes, I've been fighting this issue since last year..

**Reply by mikehardy** [reply]
That is mind-blowingly subtle + obscure. Wow. Good work getting to root cause on it! Going to file away "check expiration on refresh token" as a trick in my bag of tricks...

**Reply by MichaelHindley** [reply]
spent hours debugging the issue and this was indeed the fix(thanks for o3 research mode for finding this issue..), i was using a local emulators flow and had no way of digging into the "auth internal error"(from what i know, there is no output in the emulator ui or the debug logs for these errors), doing an explicit signout at the start of an anonymous login finally fixed it (:
